### ISSUE TITLE
Add support for Node events.on asyncIterable

### ISSUE THREAD

**Issue by filoxo** [NodeFactoryIo/fastify-sse-v2#19]
I'm pretty new to fastify, SSEs, and node events so please bear with. 

In the 'Sending events from EventEmmiters' example, it shows how to use event-iterator to convert an EventEmitter to an asyncIterable. Recently having delved into this, I found that node can already convert its EventEmitter into an asyncIterable through using [events.on](https://nodejs.org/api/events.html#events_events_on_emitter_eventname_options) which "Returns: `<AsyncIterator>` that iterates `eventName` events emitted by the `emitter`". Here is the code for a mostly working example:

```javascript
// inside a routes plugin

// contrived and simplified example
const ee = new EventEmitter();

fastify.get("/data", (_request, reply) => {
  reply.sse(on(ee, "update"));
});

fastify.post(
  "/api",
  async (request, reply) => {
    // do something...
    const update = doSomething(request.body)
    ee.emit("update", {
      type: "ping",
      data: JSON.stringify({ update }),
    });
    reply.status(200).send();
  }
);
```

Like I mentioned, this is mostly working. Where it breaks down is due to how [EventEmitters collect `emit` args into an array](https://nodejs.org/api/events.html#events_emitter_emit_eventname_args) so when [`transformAsyncIterable` in sse.ts gets a `message`](https://github.com/NodeFactoryIo/fastify-sse-v2/blob/master/src/sse.ts#L3)  from the `source`, **it is an array** not just an object like [`serializeSSEEvent`](https://github.com/NodeFactoryIo/fastify-sse-v2/blob/master/src/sse.ts#L10) expects.

So I'm opening this issue to 1) see if there's interest in supporting node `events.on` for converting to an asyncIterable, and 2) what would be the best approach for handling a message that is an array of emit arguments. 

For my needs, I'm able to assume that the first arg is all that matters:

```diff
async function* transformAsyncIterable(source) {
  for await (const message of source) {
+    if(Array.isArray(message)) {
+      yield serializeSSEEvent(message[0]);
+    }
    yield serializeSSEEvent(message);
  }

  yield serializeSSEEvent({
    event: "end",
    data: "Stream closed"
  });
}
```

but that is likely a pretty naive solution. I wonder if instead the whole array should be serialized (eg. `if(Array.isArray(message)) { yield message.map(serializeSSEEvent) }`) or if the payload should reduce them down to a single payload; I'm basing this idea off of what little I know about SSEs but the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#fields) say:

> When the `EventSource` receives multiple consecutive lines that begin with data:, it concatenates them, inserting a newline character between each one. 

So maybe it makes sense to accumulate them all and send them as `data`. I'm open to feedback and learning, thanks!

**Comment by filoxo** [comment]
I'm also going to explore node [`EventTarget`](https://nodejs.org/api/events.html#events_eventtarget_and_event_api) because it may not have this same problem.

**Comment by mpetrunic** [comment]
I'm gonna take closer look tomorrow but I guess you could do:
```javascript
reply.sse((async function*() {
   for await (const event of events.on('update')) {
        yield serializeSSEEvent({
          event: "end",
          data: "Stream closed"
      });
   }
})())
```
?

**Comment by filoxo** [comment]
`events.on` isn't the same as the `on` handler of an EventEmitter. I did try your idea though, slightly modified: 

```js
reply.sse(
  (async function* () {
    for await (const event of on(opsEvents, "update")) {
      console.log("event", event);
      yield {
        type: "ping",
        data: JSON.stringify(event),
      };
    }
  })()
);
```

so that I can attempt to transform the event however is needed before sending, but this doesn't seem to actually yield ever. Also `serializeSSEEvent` isn't exported by this plugin so it can't be used by us consumers like that.

**Comment by mpetrunic** [comment]
That's odd, it should work (I've updated my comment to be more clear). Could you provide example repo where this doesn't work?

> Also serializeSSEEvent isn't exported by this plugin so it can't be used by us consumers like that.

I didn't mean to use method from plugin, I though it was your method to convert your event to [EventMessage interface](https://github.com/NodeFactoryIo/fastify-sse-v2/blob/a36fa8c80f63793245f451c9e539ccea997eba47/src/index.ts#L13) which you should yield

**Comment by filoxo** [comment]
I created this example repo based on my original project: https://github.com/filoxo/fastify-sse-v2-events-on-asynciterable
and the relevant lines are here: https://github.com/filoxo/fastify-sse-v2-events-on-asynciterable/blob/main/routes.js#L14-L30
After removing all the additional code, I found that your idea does work! I must've gotten something wrong. Likely a mismatch of event names so it never emitted.

So, for my own understanding, the code you've provided is creating an outer async generator function so that it can iterate over the AsyncIterable returned by `events.on` so I can transform that payload into something that matches the EventMessage interface?

Oh haha sorry, I should've clarified that `serializeSSEEvent` is a function inside of this plugin, in sse.ts. 

Would this be valuable to document so that consumers can avoid one more dependency (event-iterator)? It does require node 14+ so its not for every usecase.

**Comment by mpetrunic** [comment]
> So, for my own understanding, the code you've provided is creating an outer async generator function so that it can iterate over the AsyncIterable returned by events.on so I can transform that payload into something that matches the EventMessage interface?

yes, I would also suggest to pass AbortSignal as third param and trigger it on connection closing to avoid resource leaking.
https://github.com/NodeFactoryIo/fastify-sse-v2#note

> Would this be valuable to document so that consumers can avoid one more dependency (event-iterator)? It does require node 14+ so its not for every usecase.

According to docs, seems like it works in node12 and node13: `Added in: v13.6.0, v12.16.0`
I'll add that in readme

**Comment by mpetrunic** [comment]
resolved via commit b9b77dab3ac0c400be5459cf3a508aa03aea4b79

**Comment by filoxo** [comment]
Ah, that's the correct versions of node. Thanks again for bearing with me as someone who is unfamiliar with SSEs and async iterators! This solution is great.
